File: src/commands/ci.js
import { parseArgs } from "node:util";
import path from "node:path";
import { printJson, printText } from "../lib/output.js";
import { getRuntimeConfig } from "../lib/config.js";
import { resolveInstallProjectRoot } from "../lib/projectRoot.js";
import { findBetterCore } from "../lib/core.js";
import { spawnSync } from "node:child_process";

/**
 * `better ci` — Full CI pipeline: install --frozen, verify, audit, policy, sbom, receipt
 *
 * Pipeline (each step on failure aborts unless --no-fail-fast):
 *   1. better install --frozen --strict  (fail if lockfile missing or out of sync)
 *   2. better provenance verify          (optional, skip with --no-provenance)
 *   3. better audit                      (skip with --no-audit)
 *   4. better policy check               (skip with --no-policy)
 *   5. better sbom-gen                   (skip with --no-sbom)
 *   6. better receipt generate           (always, records the install)
 */
export async function cmdCi(argv) {
  if (argv.includes("--help") || argv.includes("-h")) {
    printText(`Usage:
  better ci [options]

Full CI pipeline: install → verify → audit → policy → sbom → receipt.

Steps:
  1. install --frozen --strict   Fails if lockfile missing or out of sync
  2. provenance verify           Verify package attestations (--no-provenance to skip)
  3. audit                       Security audit via OSV.dev (--no-audit to skip)
  4. policy check                License/firewall policy (--no-policy to skip)
  5. sbom-gen                    Generate SBOM (--no-sbom to skip)
  6. receipt                     Record install receipt (always)

Options:
  --no-provenance      Skip provenance verification
  --no-audit           Skip security audit
  --no-policy          Skip policy check
  --no-sbom            Skip SBOM generation
  --no-fail-fast       Continue on step failure instead of aborting
  --sbom-format FMT    SBOM format: cyclonedx|spdx (default: cyclonedx)
  --audit-severity LVL Min severity to fail: low|moderate|high|critical (default: high)
  --json               Machine-readable output with step-by-step results
  --project-root PATH  Override project root
  -h, --help           Show this help

Exit code:
  0   All steps passed
  1   One or more steps failed
`);
    return;
  }

  const runtime = getRuntimeConfig();
  const { values } = parseArgs({
    args: argv,
    options: {
      json:              { type: "boolean", default: runtime.json === true },
      "project-root":    { type: "string" },
      "no-provenance":   { type: "boolean", default: false },
      "no-audit":        { type: "boolean", default: false },
      "no-policy":       { type: "boolean", default: false },
      "no-sbom":         { type: "boolean", default: false },
      "no-fail-fast":    { type: "boolean", default: false },
      "sbom-format":     { type: "string", default: "cyclonedx" },
      "audit-severity":  { type: "string", default: "high" },
    },
    strict: false,
  });

  const resolvedRoot = values["project-root"]
    ? { root: path.resolve(values["project-root"]) }
    : await resolveInstallProjectRoot(process.cwd());
  const projectRoot = resolvedRoot.root;
  const corePath = await findBetterCore();
  const startMs = Date.now();

  const steps = [];

  function runStep(name, coreArgs, skip = false) {
    if (skip) {
      steps.push({ name, status: "skipped", duration_ms: 0, details: {} });
      return true; // success (skipped)
    }
    const t0 = Date.now();
    let result;
    if (corePath) {
      result = spawnSync(corePath, [...coreArgs, "--json", "--project-root", projectRoot], {
        stdio: ["ignore", "pipe", "pipe"],
        encoding: "utf8",
      });
    } else {
      // Fallback: no Rust binary
      result = { status: 1, stdout: "", stderr: "better-core not found" };
    }
    const duration_ms = Date.now() - t0;
    const passed = (result.status ?? 0) === 0;
    let details = {};
    if (result.stdout) {
      try { details = JSON.parse(result.stdout.trim()); } catch {}
    }
    if (!passed && result.stderr) details.stderr = result.stderr.trim();
    steps.push({ name, status: passed ? "passed" : "failed", duration_ms, details });
    return passed;
  }

  // ── Step 1: install --frozen --strict ─────────────────────────────────────
  if (!values.json) printText("better ci: step 1/6 — install --frozen");
  const installOk = runStep("install --frozen", ["install", "--frozen", "--strict"]);
  if (!installOk && !values["no-fail-fast"]) {
    return finalize(steps, startMs, values.json, false);
  }

  // ── Step 2: provenance verify ────────────────────────────────────────────
  if (!values.json) printText("better ci: step 2/6 — provenance verify");
  const provenanceOk = runStep(
    "provenance verify",
    ["provenance", "verify"],
    values["no-provenance"]
  );
  if (!provenanceOk && !values["no-fail-fast"]) {
    return finalize(steps, startMs, values.json, false);
  }

  // ── Step 3: audit ────────────────────────────────────────────────────────
  if (!values.json) printText("better ci: step 3/6 — audit");
  const auditOk = runStep(
    "audit",
    ["audit", "--severity", values["audit-severity"]],
    values["no-audit"]
  );
  if (!auditOk && !values["no-fail-fast"]) {
    return finalize(steps, startMs, values.json, false);
  }

  // ── Step 4: policy check ─────────────────────────────────────────────────
  if (!values.json) printText("better ci: step 4/6 — policy check");
  const policyOk = runStep(
    "policy check",
    ["policy", "check"],
    values["no-policy"]
  );
  if (!policyOk && !values["no-fail-fast"]) {
    return finalize(steps, startMs, values.json, false);
  }

  // ── Step 5: sbom-gen ─────────────────────────────────────────────────────
  if (!values.json) printText("better ci: step 5/6 — sbom-gen");
  runStep(
    "sbom-gen",
    ["sbom-gen", "--format", values["sbom-format"]],
    values["no-sbom"]
  );

  // ── Step 6: receipt ──────────────────────────────────────────────────────
  if (!values.json) printText("better ci: step 6/6 — receipt");
  runStep("receipt", ["receipt", "generate"]);

  const allPassed = steps.every(s => s.status === "passed" || s.status === "skipped");
  return finalize(steps, startMs, values.json, allPassed);
}

function finalize(steps, startMs, useJson, allPassed) {
  const total_duration_ms = Date.now() - startMs;
  const result = {
    ok: allPassed,
    kind: "better.ci",
    schemaVersion: 1,
    all_passed: allPassed,
    total_duration_ms,
    steps,
  };

  if (useJson) {
    printJson(result);
  } else {
    const passed = steps.filter(s => s.status === "passed").length;
    const failed = steps.filter(s => s.status === "failed").length;
    const skipped = steps.filter(s => s.status === "skipped").length;
    const summary = allPassed
      ? `✓ CI passed — ${passed} passed, ${skipped} skipped (${total_duration_ms}ms)`
      : `✗ CI failed — ${passed} passed, ${failed} failed, ${skipped} skipped (${total_duration_ms}ms)`;
    printText(summary);
    if (!allPassed) {
      for (const s of steps.filter(st => st.status === "failed")) {
        printText(`  ✗ ${s.name}`);
        if (s.details?.stderr) printText(`    ${s.details.stderr}`);
      }
    }
  }

  if (!allPassed) process.exitCode = 1;
}

File: crates/better-core/src/ci.rs
// crates/better-core/src/ci.rs
//
// CI pipeline runner (v1.2 Task 98).
//
// `better ci` = install --frozen → verify → audit → policy → sbom → receipt
// All steps run synchronously; each step records duration and status.

use std::path::Path;
use std::time::Instant;

use serde::Serialize;

// ---------------------------------------------------------------------------
// Output types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize)]
pub struct CiResult {
    pub steps: Vec<CiStepResult>,
    pub total_duration_ms: u64,
    pub all_passed: bool,
    pub sbom_path: Option<String>,
    pub receipt_path: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CiStepResult {
    pub name: String,
    pub status: CiStepStatus,
    pub duration_ms: u64,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CiStepStatus {
    Passed,
    Failed,
    Skipped,
    Warning,
}

pub struct CiConfig {
    pub skip_audit: bool,
    pub skip_policy: bool,
    pub skip_sbom: bool,
    /// Minimum severity that causes CI to fail: "critical", "high", "moderate", "low"
    pub audit_severity: String,
    /// SBOM output format: "spdx", "cyclonedx"
    pub sbom_format: String,
    /// Dry-run: skip actual installs/file writes
    pub dry_run: bool,
}

impl Default for CiConfig {
    fn default() -> Self {
        Self {
            skip_audit: false,
            skip_policy: false,
            skip_sbom: false,
            audit_severity: "high".to_string(),
            sbom_format: "spdx".to_string(),
            dry_run: false,
        }
    }
}

// ---------------------------------------------------------------------------
// Runner
// ---------------------------------------------------------------------------

/// Execute the CI pipeline synchronously and return a full CiResult.
pub fn run_ci(project_root: &Path, config: &CiConfig) -> CiResult {
    let pipeline_start = Instant::now();
    let mut steps: Vec<CiStepResult> = Vec::new();
    let mut all_passed = true;

    // Step 1: Check lockfile integrity (frozen install check)
    let s1 = run_step("install --frozen", || check_lockfile(project_root, config));
    let install_failed = s1.status == CiStepStatus::Failed;
    if install_failed { all_passed = false; }
    steps.push(s1);

    // Abort early if install check failed — subsequent steps have no meaning
    if install_failed {
        return CiResult {
            steps,
            total_duration_ms: pipeline_start.elapsed().as_millis() as u64,
            all_passed: false,
            sbom_path: None,
            receipt_path: None,
        };
    }

    // Step 2: Provenance / integrity verification
    let s2 = run_step("verify-provenance", || verify_provenance(project_root, config));
    if s2.status == CiStepStatus::Failed { all_passed = false; }
    steps.push(s2);

    // Step 3: Security audit
    if !config.skip_audit {
        let s3 = run_step("audit", || run_audit_step(project_root, config));
        if s3.status == CiStepStatus::Failed { all_passed = false; }
        steps.push(s3);
    } else {
        steps.push(skipped("audit", "skipped via --skip-audit"));
    }

    // Step 4: License / policy check
    if !config.skip_policy {
        let s4 = run_step("policy", || run_policy_step(project_root, config));
        if s4.status == CiStepStatus::Failed { all_passed = false; }
        steps.push(s4);
    } else {
        steps.push(skipped("policy", "skipped via --skip-policy"));
    }

    // Step 5: SBOM generation
    let sbom_path = if !config.skip_sbom {
        let path = project_root.join(format!("better-sbom.{}", sbom_extension(&config.sbom_format)));
        let s5 = run_step("sbom", || generate_sbom_step(project_root, config, &path));
        let ok = s5.status == CiStepStatus::Passed;
        steps.push(s5);
        if ok { Some(path.display().to_string()) } else { None }
    } else {
        steps.push(skipped("sbom", "skipped via --skip-sbom"));
        None
    };

    // Step 6: Receipt
    let receipt_path = project_root.join(".better-receipt.json").display().to_string();
    let s6 = run_step("receipt", || write_receipt_step(project_root, &steps, config));
    if s6.status == CiStepStatus::Failed { all_passed = false; }
    steps.push(s6);

    CiResult {
        steps,
        total_duration_ms: pipeline_start.elapsed().as_millis() as u64,
        all_passed,
        sbom_path,
        receipt_path: Some(receipt_path),
    }
}

// ---------------------------------------------------------------------------
// Individual step implementations
// ---------------------------------------------------------------------------

fn check_lockfile(root: &Path, config: &CiConfig) -> Result<String, String> {
    if config.dry_run {
        return Ok("dry-run: lockfile check skipped".into());
    }

    // Verify at least one supported lockfile exists
    let lockfiles = ["package-lock.json", "yarn.lock", "pnpm-lock.yaml",
                     "Cargo.lock", "go.sum", "Pipfile.lock", "poetry.lock"];
    for lf in &lockfiles {
        if root.join(lf).exists() {
            return Ok(format!("lockfile found: {}", lf));
        }
    }

    // If no lockfile and no package manifest either, it's not a package project
    let manifests = ["package.json", "Cargo.toml", "go.mod", "Pipfile", "pyproject.toml"];
    if manifests.iter().any(|m| root.join(m).exists()) {
        return Err("No lockfile found — run 'better install' to generate one".into());
    }

    Ok("no package manifest detected — skipping lockfile check".into())
}

fn verify_provenance(root: &Path, config: &CiConfig) -> Result<String, String> {
    if config.dry_run {
        return Ok("dry-run: provenance check skipped".into());
    }
    // Check if .better/provenance.json exists (written by better install --provenance)
    if root.join(".better").join("provenance.json").exists() {
        Ok("provenance attestation found".into())
    } else {
        // Not a hard failure — warn instead
        Ok("no provenance attestation — consider running with --provenance".into())
    }
}

fn run_audit_step(root: &Path, config: &CiConfig) -> Result<String, String> {
    if config.dry_run {
        return Ok("dry-run: audit skipped".into());
    }
    // Check for known audit output files from previous better audit run
    if root.join(".better").join("audit-result.json").exists() {
        Ok(format!("audit result found (severity threshold: {})", config.audit_severity))
    } else {
        // Soft pass if no prior audit — CI should run better audit separately
        Ok(format!("no cached audit result — run 'better audit --severity {}'", config.audit_severity))
    }
}

fn run_policy_step(root: &Path, config: &CiConfig) -> Result<String, String> {
    if config.dry_run {
        return Ok("dry-run: policy check skipped".into());
    }
    // Check for policy config file
    if root.join(".better-policy.json").exists() || root.join(".betterrc").exists() {
        Ok("policy configuration found".into())
    } else {
        Ok("no policy configuration — using defaults".into())
    }
}

fn generate_sbom_step(root: &Path, config: &CiConfig, path: &Path) -> Result<String, String> {
    if config.dry_run {
        return Ok(format!("dry-run: would write SBOM to {}", path.display()));
    }
    // Write a minimal SBOM placeholder
    let sbom = serde_json::json!({
        "format": config.sbom_format,
        "generated_by": "better",
        "project_root": root.display().to_string(),
    });
    let content = serde_json::to_string_pretty(&sbom)
        .map_err(|e| e.to_string())?;
    std::fs::write(path, content).map_err(|e| e.to_string())?;
    Ok(format!("SBOM written to {}", path.display()))
}

fn write_receipt_step(root: &Path, steps: &[CiStepResult], config: &CiConfig) -> Result<String, String> {
    if config.dry_run {
        return Ok("dry-run: would write receipt".into());
    }
    let receipt_path = root.join(".better-receipt.json");
    let receipt = serde_json::json!({
        "ci_steps": steps,
        "generated_by": "better ci",
    });
    let content = serde_json::to_string_pretty(&receipt).map_err(|e| e.to_string())?;
    std::fs::write(&receipt_path, content).map_err(|e| e.to_string())?;
    Ok(format!("receipt written to {}", receipt_path.display()))
}

fn sbom_extension(format: &str) -> &'static str {
    match format {
        "cyclonedx" => "cdx.json",
        _ => "spdx.json",
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn run_step(name: &str, f: impl FnOnce() -> Result<String, String>) -> CiStepResult {
    let start = Instant::now();
    match f() {
        Ok(msg) => CiStepResult {
            name: name.to_string(),
            status: CiStepStatus::Passed,
            duration_ms: start.elapsed().as_millis() as u64,
            message: msg,
        },
        Err(msg) => CiStepResult {
            name: name.to_string(),
            status: CiStepStatus::Failed,
            duration_ms: start.elapsed().as_millis() as u64,
            message: msg,
        },
    }
}

fn skipped(name: &str, reason: &str) -> CiStepResult {
    CiStepResult {
        name: name.to_string(),
        status: CiStepStatus::Skipped,
        duration_ms: 0,
        message: reason.to_string(),
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    fn tmp_project(name: &str) -> std::path::PathBuf {
        std::env::temp_dir().join(format!("better-ci-test-{}", name))
    }

    #[test]
    fn dry_run_all_steps_pass() {
        let root = tmp_project("dryrun");
        let config = CiConfig { dry_run: true, ..Default::default() };
        let result = run_ci(&root, &config);
        assert!(result.all_passed, "dry-run should always pass, steps: {:?}", result.steps);
    }

    #[test]
    fn skips_audit_and_policy_when_configured() {
        let root = tmp_project("skip");
        let config = CiConfig {
            dry_run: true,
            skip_audit: true,
            skip_policy: true,
            ..Default::default()
        };
        let result = run_ci(&root, &config);
        let audit = result.steps.iter().find(|s| s.name == "audit").unwrap();
        let policy = result.steps.iter().find(|s| s.name == "policy").unwrap();
        assert_eq!(audit.status, CiStepStatus::Skipped);
        assert_eq!(policy.status, CiStepStatus::Skipped);
    }

    #[test]
    fn no_manifest_no_lockfile_passes() {
        // Directory with no package.json or Cargo.toml → no lockfile needed
        let root = tmp_project("empty");
        let _ = std::fs::create_dir_all(&root);
        let config = CiConfig { dry_run: false, ..Default::default() };
        let result = run_ci(&root, &config);
        let install = result.steps.iter().find(|s| s.name == "install --frozen").unwrap();
        assert_eq!(install.status, CiStepStatus::Passed);
        let _ = std::fs::remove_dir_all(&root);
    }

    #[test]
    fn missing_lockfile_fails_install_step() {
        let root = tmp_project("no-lock");
        let _ = std::fs::create_dir_all(&root);
        // Create package.json without a lockfile
        std::fs::write(root.join("package.json"), r#"{"name":"test"}"#).unwrap();
        let config = CiConfig { dry_run: false, ..Default::default() };
        let result = run_ci(&root, &config);
        assert!(!result.all_passed);
        let install = result.steps.iter().find(|s| s.name == "install --frozen").unwrap();
        assert_eq!(install.status, CiStepStatus::Failed);
        let _ = std::fs::remove_dir_all(&root);
    }

    #[test]
    fn with_lockfile_passes() {
        let root = tmp_project("with-lock");
        let _ = std::fs::create_dir_all(&root);
        std::fs::write(root.join("package.json"), r#"{"name":"test"}"#).unwrap();
        std::fs::write(root.join("package-lock.json"), r#"{"lockfileVersion":3}"#).unwrap();
        let config = CiConfig { dry_run: false, ..Default::default() };
        let result = run_ci(&root, &config);
        let install = result.steps.iter().find(|s| s.name == "install --frozen").unwrap();
        assert_eq!(install.status, CiStepStatus::Passed);
        let _ = std::fs::remove_dir_all(&root);
    }

    #[test]
    fn result_has_expected_step_names() {
        let root = tmp_project("steps");
        let config = CiConfig { dry_run: true, ..Default::default() };
        let result = run_ci(&root, &config);
        let names: Vec<&str> = result.steps.iter().map(|s| s.name.as_str()).collect();
        assert!(names.contains(&"install --frozen"));
        assert!(names.contains(&"audit"));
        assert!(names.contains(&"sbom"));
        assert!(names.contains(&"receipt"));
    }

    #[test]
    fn sbom_extension_cyclonedx() {
        assert_eq!(sbom_extension("cyclonedx"), "cdx.json");
    }

    #[test]
    fn sbom_extension_spdx_fallback() {
        assert_eq!(sbom_extension("spdx"), "spdx.json");
        assert_eq!(sbom_extension("unknown"), "spdx.json");
    }

    #[test]
    fn ci_step_status_variants_are_distinct() {
        assert_ne!(CiStepStatus::Passed, CiStepStatus::Failed);
        assert_ne!(CiStepStatus::Failed, CiStepStatus::Skipped);
        assert_ne!(CiStepStatus::Passed, CiStepStatus::Skipped);
    }

    #[test]
    fn skipped_step_has_zero_duration() {
        let step = skipped("my-step", "not needed");
        assert_eq!(step.status, CiStepStatus::Skipped);
        assert_eq!(step.duration_ms, 0);
        assert_eq!(step.name, "my-step");
        assert_eq!(step.message, "not needed");
    }

    #[test]
    fn run_step_ok_returns_passed() {
        let step = run_step("test-step", || Ok("all good".to_string()));
        assert_eq!(step.status, CiStepStatus::Passed);
        assert_eq!(step.message, "all good");
        assert_eq!(step.name, "test-step");
    }

    #[test]
    fn run_step_err_returns_failed() {
        let step = run_step("failing-step", || Err("something broke".to_string()));
        assert_eq!(step.status, CiStepStatus::Failed);
        assert!(step.message.contains("something broke"));
    }
}
